common/cf-guest/src/client_impls.rs

```diff
@@ -369,26 +394,95 @@ impl<PK: PubKey> ClientState<PK> {
 
     fn verify_misbehaviour(
         &self,
-        _ctx: &impl guestchain::Verifier<PK>,
-        _misbehaviour: Misbehaviour<PK>,
+        ctx: &impl guestchain::Verifier<PK>,
+        misbehaviour: Misbehaviour<PK>,
     ) -> Result<()> {
-        todo!()
+        let Misbehaviour { header1, header2 } = misbehaviour;
+
+        // If the headers belong to different chains, they aren’t proof of
+        // misbehaviour.
+        if header1.genesis_hash != header2.genesis_hash {
+            return Err(error("Headers belong to different blockchains"));
+        }
+
+        self.verify_header(ctx, header1)?;
+        self.verify_header(ctx, header2)?;
+
+        Ok(())
     }
 
-    fn check_for_misbehaviour_header(
+    fn check_for_misbehaviour_in_header(
         &self,
-        _ctx: &impl guestchain::Verifier<PK>,
-        _header: Header<PK>,
+        ctx: &impl CommonContext,
+        client_id: &ibc::ClientId,
+        header: Header<PK>,
     ) -> Result<bool> {
-        todo!()
+        fn get_timestamp_ns<T, E>(state: Option<T>) -> Result<Option<u64>>
+        where
+            E: ToString,
+            T: TryInto<ConsensusState, Error = E>,
+        {
+            match state.map(|state| state.try_into()) {
+                None => Ok(None),
+                Some(Ok(state)) => Ok(Some(state.timestamp_ns.get())),
+                Some(Err(err)) => Err(error(err)),
+            }
+        }
+
+        let height = header.block_header.block_height;
+        let height = ibc::Height::new(1, height.into())?;
+
+        Ok(match ctx.consensus_state_neighbourhood(client_id, height)? {
+            Neighbourhood::This(state) => {
+                // If we already have existing consensus for given height, check
+                // that what we’ve been sent is the same thing we have.  If it
+                // isn’t, that’s evidence of misbehaviour.
+                let existing_state = state.try_into().map_err(error)?;
+                let header_state = ConsensusState::from(&header);
+                existing_state != header_state
+            }
+
+            Neighbourhood::Neighbours(prev, next) => {
+                // Otherwise, make sure that timestamp of each consensus is
+                // strictly increasing.  If it isn’t, that’s evidence of
+                // misbehaviour.
+                let header_time_ns = header.block_header.timestamp_ns.get();
+                if let Some(prev_time_ns) = get_timestamp_ns(prev)? {
+                    if header_time_ns <= prev_time_ns {
```

Comment:
can it be equal if it's the previous?

Reply:
No.  Timestamps are strictly increasing.  This is checked in Block::generate_next:

```rust
        } else if timestamp_ns <= self.timestamp_ns {
            return Err(GenerateError::BadHostTimestamp);
        }
```

(here `self` is the current block and `timestamp_ns` is timestamp of
the next block to be generated).